superdesk/websockets_comms.py

```diff
@@ -192,6 +204,52 @@ def _client_loop(self, websocket):
             pings += 1
             yield from websocket.send(json.dumps({"ping": pings, "clients": len(websocket.ws_server.websockets)}))
 
+    def get_message_recipients(self, message_data: WebsocketMessageData) -> Set[WebSocketClientProtocol]:
+        """Filter consumers by message filter attributes
+
+        When client's connect, they can provide a set of URL arguments as to what they need to subscribe to
+        Such as specific user and/or company combination.
+        These URL arguments are then checked against the filter conditions in the ``message_data``
+        to determine which clients this ``message_data`` is to be sent to
+        """
+
+        clients = self.clients.copy()
+
+        if not message_data.get("filters"):
+            return clients
+
+        filters: WebsocketMessageFilterConditions = message_data.pop("filters", {})
+        filters.setdefault("include", {})
+        filters.setdefault("exclude", {})
+
+        if not filters["include"] and not filters["exclude"]:
+            return clients
+
+        def _filter(websocket: WebSocketClientProtocol) -> bool:
+            if filters["include"] and not websocket.path.startswith(self.subscribe_prefix):
+                # If ``filter.include`` is defined, client must provide url args in websocket path
+                # as we're explicitly including only clients that have args in this list
+                return False
+
+            try:
+                url_args: Dict[str, str] = {
```

Comment:
I suppose it's not a big deal, but this could be done once when client connects and not every time there is a new message

Reply:
Good optimisation, I'll do that tomorrow (as to keep ws service running as fast as possible)